cache.go

```diff
@@ -0,0 +1,152 @@
+func newStatementCache() *statementCache {
+	once.Do(func() {
+		singleStmtCache = &statementCache{
+			stmtDBCache: map[uint64]map[uint64]*sql.Stmt{},
+			dbStmtCache: map[uint64]map[uint64]bool{},
+		}
+	})
+	return singleStmtCache
+}
+
+// newStatement returns a new sqlair.Statement and adds it to the cache. A
+// finalizer is set on the sqlair.Statement to remove its ID from the cache and
+// close all associated sql.Stmt objects.
+func (sc *statementCache) newStatement(te *expr.TypedExpr) *Statement {
+	cacheID := atomic.AddUint64(&sc.stmtIDCount, 1)
+	sc.mutex.Lock()
+	sc.stmtDBCache[cacheID] = map[uint64]*sql.Stmt{}
+	sc.mutex.Unlock()
+	s := &Statement{te: te, cacheID: cacheID}
+	// This finalizer is run after the Statement is garbage collected.
+	runtime.SetFinalizer(s, sc.getStmtFinalizer(s))
+	return s
+}
+
+// newDB returns a new sqlair.DB and allocates the necessary resources in the
+// statementCache. A finalizer is set on the sqlair.DB to remove references to
+// it from the cache, close all sql.Stmt objects on it, and close the sql.DB
+func (sc *statementCache) newDB(sqldb *sql.DB) *DB {
+	cacheID := atomic.AddUint64(&sc.dbIDCount, 1)
+	sc.mutex.Lock()
+	sc.dbStmtCache[cacheID] = map[uint64]bool{}
+	sc.mutex.Unlock()
+	db := &DB{sqldb: sqldb, cacheID: cacheID}
+	// This finalizer is run after the DB is garbage collected.
+	runtime.SetFinalizer(db, sc.getDBFinalizer(db))
+	return db
+}
+
+// prepareSubstrate is an object that queries can be prepared on, e.g. a sql.DB
+// or sql.Conn. It is used in prepareStmt.
+type prepareSubstrate interface {
+	PrepareContext(context.Context, string) (*sql.Stmt, error)
+}
+
+// prepareStmt prepares a Statement on a prepareSubstrate. It first checks in
+// the cache to see if it has already been prepared on the DB.
+// The prepareSubstrate must be associated with the same DB that prepareStmt is
+// a method of.
+func (sc *statementCache) prepareStmt(ctx context.Context, dbID uint64, ps prepareSubstrate, s *Statement) (*sql.Stmt, error) {
+	var err error
+	sc.mutex.RLock()
+	// The Statement cache ID is only removed from stmtDBCache when the
+	// finalizer is run. The cache ID must be present as we have a reference to
+	// the Statement.
+	sqlstmt, ok := sc.stmtDBCache[s.cacheID][dbID]
+	sc.mutex.RUnlock()
+	if !ok {
+		sqlstmt, err = ps.PrepareContext(ctx, s.te.SQL())
+		if err != nil {
+			return nil, err
+		}
+		sc.mutex.Lock()
+		// Check if a statement has been inserted by someone else since we last
+		// checked.
+		sqlstmtAlt, ok := sc.stmtDBCache[s.cacheID][dbID]
+		if ok {
+			sqlstmt.Close()
+			sqlstmt = sqlstmtAlt
+		} else {
+			sc.stmtDBCache[s.cacheID][dbID] = sqlstmt
+			sc.dbStmtCache[dbID][s.cacheID] = true
+		}
+		sc.mutex.Unlock()
+	}
+	return sqlstmt, nil
+}
+
+// getStmtFinalizer returns a finalizer that removes a Statement from the
+// statement caches and closes it.
+func (sc *statementCache) getStmtFinalizer(s *Statement) func(*Statement) {
```

Comment:
Can we rename this function? I think that the name does not add anything now because it is talking about its usage and, said usage already tells us this is a finalizer:
```go
runtime.SetFinalizer(s, sc.getStmtFinalizer(s))
                        ----------------------
                        we already know this is a finalizer based on the function signature
```
Can we instead rename it to be `cleanCacheAndCloseStmtThunk`, if it is too verbose then `closeStmtInCacheThunk`, whatever conveys what is does instead of what it is used for.

Reply:
Fair, I've renamed these functions. I changed it to "Func" instead of "Thunk" because I'm not sure it'd be entirely clear to readers of the code what Thunk meant in this context.

Reply:
Well thunk seems to have a more defined meaning (https://en.wikipedia.org/wiki/Thunk) but I like `Func` as well.